### ISSUE TITLE
[Feature] dbt failure scenarios need more nuance

### ISSUE THREAD

**Issue by mirosval** [dbt-labs/dbt-core#5339]
### Is this your first time opening an issue?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)

### Context
I'm running dbt with Airflow such that we schedule individual model runs and related tests in a single AirFlow Task. I would like the Airflow task status to reflect the result of dbt execution. 

### The Problem
Presently dbt does not distinguish among different failure scenarios, in particular there is no way to distinguish between DB-related errors and dbt test result errors. These are 2 types of errors that belong to completely distinct domains and need to be treated separately.

In particular, I would like to:
* Fail the Airflow task when there are dbt selector or database errors
* Not fail the Airflow task when there are dbt test failures (instead I just note the failures and let the ETL continue)

### How it differs from #3369
There is a workaround suggested in https://github.com/dbt-labs/dbt-core/issues/3369 that has worked for me partially, which is to use `dbt list --select <models>` and let the task fail if this fails.

This _almost_ works, it works for the cases when there is an erroneous selector specified, task doesn't exist or there are problems with how the task is defined (bad refs in dbt, etc). This is very good already, but it fails in one specific case and that is when there are **transient issues with the DB**. The list task can succeed if everything is fine with the model, but then the `dbt test` task will fail later because the db connection pool is exhausted.

### Proposal
Introduce a flag to dbt that separates these 2 error domains. I would like to:
* keep test failures reported as errors, when their severity is not downgraded to `warn`
* still have the `dbt test` or `dbt source freshness` fail if there are database errors

### Describe alternatives you've considered

I'm already mitigating this problem with `dbt list` which helps in some cases, but as indicated above, its not enough.

The only other alternative for me is to parse the log output of the dbt task which is very messy and I would rather not

### Who will this benefit?

Anyone who runs dbt using their own scheduling

### Are you interested in contributing this feature?

yes

### Anything else?

I have previously filed this issue under https://github.com/dbt-labs/dbt-core/issues/3369 but since I can't reopen that one I decided to post a follow-up.

**Comment by mirosval** [comment]
Thanks, generally I think this makes sense, however it should be noted that (in some cases?) dbt does _not_ produce `run_results.json`. I guess it happens when you have DB timeouts or invalid credentials or something that prevents you from connecting to the db in the first place.

**Comment by jtcohen6** [comment]
@mirosval That's true. If dbt fails to establish any connection to the warehouse in the first place (e.g. no internet connection), it won't produce `run_results.json`, and it will emit the [`MainEncounteredError`](https://github.com/dbt-labs/dbt-core/blob/16dc2be556026674978ec0b61dadf2c1a128cd1f/core/dbt/events/types.py#L105) event.
